2D-DCT

mkatsimpris opened the pull request:
@cfelton, @jck, @josyb, @dargor0, @nkkav

cfelton committed:
Merge pull request #13 from mkatsimpris/master

Color space conversion module and testbench

Commit:
2d_dct numpy

Commit:
2d_dct changes

Commit:
1d_dct module implemented, and 2d_dct software prototype

Commit:
fixed an error in 1d-dct, now it works fine

mkatsimpris commented:
The code in the branch is in the early stage of development.

cfelton commented:
A very good start!

cfelton commented on jpegenc/subblocks/2d_dct/test_1d_dct.py:
```diff
@@ -0,0 +1,73 @@
+               0xe9, 0xe0, 0xf1, 0xff, 0xef, 0xad, 0x8a, 0x90,
+               0xe7, 0xf2, 0xf1, 0xeb, 0xf7, 0xfb, 0xd0, 0x97]
+
+@myhdl.block
```
the `clock_driver` and the `reset_on_start` should be put in a separate module and imported in the tests, since it is used in more than one test.  In the test directory there is a local package `support`, you could put them somewhere in the `support` package.

The usage in the test modules would be

``` python
from support import clock_driver, reset_on_start
```

mkatsimpris replied:
Of course there is code which is used by other modules and it should be in file. Also some interfaces must be in the same file.

mkatsimpris replied:
The code is in early stage of development and experimentation. A lot of things will be changed the following days.

mkatsimpris committed:
created 2d-dct module and test

mkatsimpris committed:
added interfaces

mkatsimpris committed:
changes

Commit:
subtracter 128 from inputs, 2d-dct outputs correct results

mkatsimpris committed:
correction of output_2d data_valid signal

mkatsimpris committed:
added new directory for dct_1d and refactore test unit

Commit:
changes test_dct_1d

mkatsimpris committed:
minor changes

mkatsimpris committed:
test_dct_1d works perfect

Commit:
2d-dct testbench completed

cfelton commented on jpegenc/subblocks/dct_1d/interfaces.py:
```diff
@@ -0,0 +1,117 @@
+        self.y74 = Signal(intbv(0, min=-out_range, max=out_range))
+        self.y75 = Signal(intbv(0, min=-out_range, max=out_range))
+        self.y76 = Signal(intbv(0, min=-out_range, max=out_range))
+        self.y77 = Signal(intbv(0, min=-out_range, max=out_range))
```
First, why not use a list-of-signals?

If an LoS can't be used, there is a little python trick you can do to build the discrete outputs.

```
class outputs(object):
    def __init__(self):
        self._build_output_attributes()
    def _build_output_attributes(self):
        # 80 should be an argument (modular / scalable)
        for ii in range(80):
            self.__dict__['y{:02d}'.format(ii)] = Signal(intbv(0, min=-8, max=8))

outs = outputs()
pprint(vars(outs))

{'y00': Signal(intbv(0)),
 'y01': Signal(intbv(0)),
 'y02': Signal(intbv(0)),
 'y03': Signal(intbv(0)),
 'y04': Signal(intbv(0)),
 'y05': Signal(intbv(0)),
 'y06': Signal(intbv(0)),
 'y07': Signal(intbv(0)),
 'y08': Signal(intbv(0)),
 'y09': Signal(intbv(0)),
 'y10': Signal(intbv(0)),
 'y11': Signal(intbv(0))
...
```

mkatsimpris commented:
I have tried to use list of signals in the interface but it is not convertible. I will try the python trick to see if it works.

mkatsimpris commented on jpegenc/subblocks/dct_2d/dct_2d.py:
```diff
@@ -0,0 +1,232 @@
+        outputs.y11.next = outputs_2nd_stage_1.out1
+        outputs.y12.next = outputs_2nd_stage_2.out1
+        outputs.y13.next = outputs_2nd_stage_3.out1
+        outputs.y14.next = outputs_2nd_stage_4.out1
```
@cfelton is there any way to avoid these assignments with less code?
I tried setattr() but it is not convertible

mkatsimpris committed:
outputs_2d changed

mkatsimpris committed:
added the tests in test directory

mkatsimpris committed:
fix a print in test_dct_2d.py

mkatsimpris committed:
minor changes

mkatsimpris committed:
minor changes

mkatsimpris committed:
added __init__.py

mkatsimpris committed:
removes redundant files

mkatsimpris committed:
removed python 3.4 and 3.5 from travis.yml as there are problem in conversion with tuples

mkatsimpris committed:
added a bug

Commit:
minor changes

Commit:
minor changes

mkatsimpris committed:
docstrings added

mkatsimpris committed:
minor changes

cfelton commented on jpegenc/subblocks/dct_1d/dct_1d.py:
```diff
@@ -128,23 +128,23 @@ def coeff_assign():
-                adder_reg[5].next = mux_flush[5] + mult_reg[5]
-                adder_reg[6].next = mux_flush[6] + mult_reg[6]
-                adder_reg[7].next = mux_flush[7] + mult_reg[7]
+            mult_reg[0].next = data_in_reg * coeffs[0]
```
Is there a reason why a loop is not used?

``` python
for ii in range(8):
    mult_reg[ii].next = data_in_reg * coeffs[ii]
```

mkatsimpris replied:
Yes there is a reason. If a for loop is used then none of the Xilinx or Altera synthesizers will recognize the multiplications as DSP units. Instead with this configuration all the multiplications are recognized and assigned to DSP units. Check page 99 of http://www.xilinx.com/itp/xilinx10/books/docs/sim/sim.pdf, page 6 of http://ridl.cfd.rit.edu/products/manuals/altera/qts_qii51007.pdf

cfelton replied:
I don't recall the being the situations (I think I have some counter examples) but I will need to search for them.

If device specific optimization are implemented these need to be explicitly stated (comments, docstrings, etc.).  Personally, I don't like premature optimization, I would rather see a clean and a modular (as possible) implementation, then get some data on resource usage, and then iterate (figure out how to include the optimizations if needed).

cfelton replied:
Here is an example where DSP48s (`mac` in the figure) are inferred fine from a loop.  This shows the percentage of resources used on a spartan 6 for a pipelined parallel FIR filter.  It is implemented with loops (no way in hell I would gather all this data having to manually change the code for the filter sizes).

![dsp48_usage_example](https://cloud.githubusercontent.com/assets/766391/16303530/b13f0050-3916-11e6-820d-ef32257cd90a.png)

The inference depends less if a loop is used or not but more on the number of registers before and after the operation.  Being fairly familiar with the synthesis guidelines I don't recall any limitation outlined and I didn't see them in the pages you referred.  What am I missing?

In this project we should be able to generate similar data, for the transform implementation what are the resource utilization for different settings.

mkatsimpris replied:
Thanks Chris for the data that you provided. I will change the code to for loops.

cfelton commented on jpegenc/subblocks/dct_2d/dct_2d.py:
```diff
@@ -0,0 +1,254 @@
+
+
+@myhdl.block
+def dct_2d(inputs, outputs, clock, reset, num_fractional_bits=14, out_prec=10):
```
@mkatsimpris can you review [this example](https://gist.github.com/cfelton/595fdf0ca1b8c61264397e37da6d938f) and let me know if a generic interface, like the one in the linked example, will work for your inputs and outputs.

If so, then we should investigate if we can encapsulate the 2D blocks even more, maybe doing [something like this](https://gist.github.com/cfelton/f6a1a1e0cfb2204cd2ee) and [modify this](https://github.com/cfelton/test_jpeg/blob/master/jpegenc/models/buffers/block_buffer.py)

The goal is to make it as generic, modular and readable as possible.

mkatsimpris replied:
I will check it out Chris.

mkatsimpris committed:
minor changes

Vikram9866 committed:
The initial implementation of the run-length-encoder (RLE)

This is the initial implementation of the run-length-encoder.  Some integration work, documentation, and refactoring needs to be completed.  The package and subpackages need to be streamlined so that it is straightforward to access the public API.  The public interfaces need to placed in a common place so that it is easy to build the system from the interfaces.  The interfaces all need to be merged, there should be a small number of global interfaces used in this design (any number of local interfaces can be used).

cfelton commented:
@mkatsimpris heads up, I messed things up in the main repo.  I tried merging some changes and didn't notice some files were removed in the pull-request.  Don't pull from the main repo anytime soon.  Will let you know when I have it cleaned-up.

cfelton committed:
Revert "added rle files"

cfelton committed:
Merge pull request #15 from cfelton/revert-14-rle

Revert "added rle files"

mkatsimpris committed:
dct_1d more modular and scalable in progress

mkatsimpris committed:
major changes

mkatsimpris committed:
minor changes

Commit:
dct_1d refactored for N vectors

Commit:
2d-dct support NxN blocks of data

Commit:
minor changes

cfelton committed:
Update romr.py

Vikram9866 committed:
The initial implementation of the run-length-encoder (RLE) (reworked)

This is the initial implementation of the run-length-encoder.  Some integration work, documentation, and refactoring needs to be completed.  The package and subpackages need to be streamlined so that it is straightforward to access the public API.  The public interfaces need to placed in a common place so that it is easy to build the system from the interfaces.  The interfaces all need to be merged, there should be a small number of global interfaces used in this design (any number of local interfaces can be used).

* Update rlecore.py
* Update rlecore.py
* modified entropycoder
* updated rlecore
* added cfelton/rhea in requirements.txt
...

cfelton commented on jpegenc/subblocks/dct_1d/dct_1d.py:
```diff
@@ -0,0 +1,205 @@
+        """rounding"""
+        if cycles_counter == N + 2 or (first_row_passed and
+                                       cycles_counter == N - 1):
+            for i in range(N):
```
Does the for loop need to be in the conditional?  I think only
the valid does, this should reduce the logic in the datapath.

mkatsimpris replied:
Yes you are right Chris!

cfelton commented:
@mkatsimpris looking good, the "generalized" implementation will be very useful.  Is the transform limited to squared blocks?  Some references mention non-squared block performance, like a 16x8 or other.  Is it reasonable to consider non square blocks?

I have some other minor comments but I want to review them before commenting.  I expect changes to be made but you can create a PR to the main repo if you want.

mkatsimpris commented:
@cfelton The particular method which used in the implementation (row collumn decomposition) supports only NxN blocks.

Commit:
minor changes

Commit:
minor changes

Vikram9866 committed:
Signal data type inheriting fix

Commit:
minor changes

mkatsimpris committed:
minor changes
